libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,995 @@
+#include <libsolidity/lsp/ReferenceCollector.h>
+
+#include <liblangutil/SourceReferenceExtractor.h>
+#include <liblangutil/CharStream.h>
+
+#include <libsolutil/Visitor.h>
+#include <libsolutil/JSON.h>
+
+#include <boost/filesystem.hpp>
+#include <boost/algorithm/string/predicate.hpp>
+
+#include <fmt/format.h>
+
+#include <ostream>
+#include <string>
+
+using namespace std;
+using namespace std::placeholders;
+
+using namespace solidity::langutil;
+using namespace solidity::frontend;
+
+namespace solidity::lsp {
+
+namespace // {{{ helpers
+{
+
+string toFileURI(boost::filesystem::path const& _path)
+{
+	return "file://" + _path.generic_string();
+}
+
+class ASTNodeLocator: public ASTConstVisitor
+{
+public:
+	static ASTNode const* locate(int _pos, SourceUnit const& _sourceUnit)
+	{
+		ASTNodeLocator locator{_pos};
+		_sourceUnit.accept(locator);
+		return locator.m_closestMatch;
+	}
+
+	bool visitNode(ASTNode const& _node) override
+	{
+		if (_node.location().contains(m_pos))
+		{
+			m_closestMatch = &_node;
+			return true;
+		}
+		return false;
+	}
+
+private:
+	explicit ASTNodeLocator(int _pos): m_pos{_pos}, m_closestMatch{nullptr} {}
+
+	int m_pos;
+	ASTNode const* m_closestMatch;
+};
+
+optional<string> extractPathFromFileURI(std::string const& _uri)
+{
+	if (!boost::starts_with(_uri, "file://"))
+		return nullopt;
+
+	return _uri.substr(7);
+}
+
+Json::Value toJson(LineColumn _pos)
+{
+	Json::Value json = Json::objectValue;
+	json["line"] = max(_pos.line, 0);
+	json["character"] = max(_pos.column, 0);
+
+	return json;
+}
+
+Json::Value toJsonRange(int _startLine, int _startColumn, int _endLine, int _endColumn)
+{
+	Json::Value json;
+	json["start"] = toJson({_startLine, _startColumn});
+	json["end"] = toJson({_endLine, _endColumn});
+	return json;
+}
+
+Json::Value toJsonRange(SourceLocation const& _location)
+{
+	solAssert(_location.source, "");
+	auto const [startLine, startColumn] = _location.source->translatePositionToLineColumn(_location.start);
+	auto const [endLine, endColumn] = _location.source->translatePositionToLineColumn(_location.end);
+	return toJsonRange(startLine, startColumn, endLine, endColumn);
+}
+
+Json::Value toJson(boost::filesystem::path const& _basePath, SourceLocation const& _location)
+{
+	solAssert(_location.source.get() != nullptr, "");
+	Json::Value item = Json::objectValue;
+	item["uri"] = toFileURI(_basePath / boost::filesystem::path(_location.source->name()));
+	item["range"] = toJsonRange(_location);
+	return item;
+}
+
+std::pair<size_t, size_t> offsetsOf(std::string const& _text, LineColumnRange _range) noexcept
+{
+	auto const start = CharStream::translateLineColumnToPosition(_text, _range.start.line, _range.start.column);
+	auto const end = CharStream::translateLineColumnToPosition(_text, _range.end.line, _range.end.column);
+	solAssert(start.has_value(), "");
+	solAssert(end.has_value(), "");
+	return std::pair{ size_t(start.value()), size_t(end.value()) };
```

Comment:
```suggestion
	return std::pair{size_t(start.value()), size_t(end.value())};
```

Reply:
Can you convert to `size_t` without `static_cast`?

Reply:
yes. `static_cast` is equivalent to a function-call style cast. (IIRC)